src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringFailoverPluginTest.java

```diff
@@ -0,0 +1,312 @@
+    Mockito
+        .when(nativeFailureDetectionEnabledProperty.getValue())
+        .thenReturn(Boolean.TRUE);
+    initializePlugin();
+
+    Assertions.assertDoesNotThrow(() -> {
+      plugin.execute(NO_MONITOR_METHOD_NAME, sqlFunc);
+      Mockito
+          .verify(mockPlugin)
+          .execute(Mockito.eq(NO_MONITOR_METHOD_NAME), Mockito.eq(sqlFunc));
+    });
+  }
+
+  @Test
+  void test_6_executeThrowsExecutionException() {
+    Mockito
+        .when(nativeFailureDetectionEnabledProperty.getValue())
+        .thenReturn(Boolean.TRUE);
+
+    initializePlugin();
+
+    Assertions.assertThrows(Exception.class, () -> {
+      Mockito
+          .when(mockPlugin.execute(Mockito.any(), Mockito.any()))
+          .thenThrow(EXECUTION_EXCEPTION);
+
+      plugin.execute(MONITOR_METHOD_NAME, sqlFunc);
+    });
+
+    Mockito.verify(monitorService).stopMonitoring(Mockito.eq(context));
+    Mockito.verify(logger, Mockito.atLeastOnce()).logTrace(Mockito.anyString());
+  }
+
+  @Test
+  void test_7_executeWithUnhealthyNode() {
+    Mockito
+        .when(nativeFailureDetectionEnabledProperty.getValue())
+        .thenReturn(Boolean.TRUE);
+
+    initializePlugin();
+
+    Assertions.assertThrows(CJCommunicationsException.class, () -> {
+      Mockito
+          .when(context.isNodeUnhealthy())
+          .thenReturn(Boolean.TRUE);
+
+      Mockito
+          .when(mockPlugin.execute(Mockito.any(), Mockito.any()))
+          .thenAnswer(invocation -> {
+            // Imitate running a long query;
+            Thread.sleep(60 * 1000);
+            return "done";
+          });
+
+      plugin.execute(MONITOR_METHOD_NAME, sqlFunc);
+    });
+
+    Mockito.verify(context).isNodeUnhealthy();
+    Mockito.verify(monitorService).stopMonitoring(Mockito.eq(context));
+    Mockito.verify(logger, Mockito.atLeastOnce()).logTrace(Mockito.anyString());
+  }
+
+  @Test
+  void test_8_executeSuccessfulWithLongQuery() {
+    final String expected = "foo";
+
+    Mockito
+        .when(nativeFailureDetectionEnabledProperty.getValue())
+        .thenReturn(Boolean.TRUE);
+
+    initializePlugin();
+
+    Assertions.assertDoesNotThrow(() -> {
+      Mockito
+          .when(context.isNodeUnhealthy())
+          .thenReturn(Boolean.FALSE);
+
+      Mockito
+          .when(mockPlugin.execute(Mockito.any(), Mockito.any()))
+          .thenAnswer(invocation -> {
+            // Imitate running a query for 10 seconds.
+            Thread.sleep(10 * 1000);
+            return expected;
+          });
+
+      final Object result = plugin.execute(MONITOR_METHOD_NAME, sqlFunc);
+      Assertions.assertEquals(expected, result);
+    });
+
+    Mockito.verify(context, Mockito.atLeastOnce()).isNodeUnhealthy();
+    Mockito.verify(monitorService).stopMonitoring(Mockito.eq(context));
+    Mockito.verify(logger, Mockito.atLeastOnce()).logTrace(Mockito.anyString());
+  }
+
+  private static Stream<Arguments> generateNullArguments() {
```

Comment:
The code is easy to read, but i would suggest either adding a comment mentioning you are creating multiple objects with one single null argument